source/agent/addons/audioRanker/AudioRankerWrapper.cc

```diff
@@ -0,0 +1,151 @@
+  }
+
+  if (info.IsConstructCall()) {
+    AudioRanker* obj = new AudioRanker();
+
+    obj->asyncResource_ = new Nan::AsyncResource("RankChange");
+    obj->callback_ = new Nan::Callback(info[0].As<Function>());
+    uv_async_init(uv_default_loop(), &obj->async_, &AudioRanker::Callback);
+
+    if (info.Length() > 2) {
+        bool detectMute = Nan::To<bool>(info[1]).FromJust();
+        int changeInterval = Nan::To<int>(info[2]).FromJust();
+        obj->me = new owt_base::AudioRanker(obj, detectMute, changeInterval);
+    } else {
+        obj->me = new owt_base::AudioRanker(obj);
+    }
+
+    obj->Wrap(info.This());
+    info.GetReturnValue().Set(info.This());
+  } else {
+    // const int argc = 1;
+    // v8::Local<v8::Value> argv[argc] = {info[0]};
+    // v8::Local<v8::Function> cons = Nan::New(constructor);
+    // info.GetReturnValue().Set(cons->NewInstance(argc, argv));
+  }
+}
+
+NAN_METHOD(AudioRanker::close) {
+  AudioRanker* obj = Nan::ObjectWrap::Unwrap<AudioRanker>(info.Holder());
+
+  if (!uv_is_closing(reinterpret_cast<uv_handle_t*>(&obj->async_))) {
+    uv_close(reinterpret_cast<uv_handle_t*>(&obj->async_), NULL);
+  }
+
+  delete obj->me;
+  obj->me = nullptr;
+  delete obj->callback_;
+  delete obj->asyncResource_;
+}
+
+NAN_METHOD(AudioRanker::addOutput) {
+  AudioRanker* obj = Nan::ObjectWrap::Unwrap<AudioRanker>(info.Holder());
+  owt_base::AudioRanker* me = obj->me;
+
+  FrameDestination* param = node::ObjectWrap::Unwrap<FrameDestination>(info[0]->ToObject());
+  owt_base::FrameDestination* dest = param->dest;
+
+  me->addOutput(dest);
+}
+
+NAN_METHOD(AudioRanker::addInput) {
+  AudioRanker* obj = Nan::ObjectWrap::Unwrap<AudioRanker>(info.Holder());
+  owt_base::AudioRanker* me = obj->me;
+
+  FrameSource* param = node::ObjectWrap::Unwrap<FrameSource>(info[0]->ToObject());
+  owt_base::FrameSource* src = param->src;
+
+  Nan::Utf8String streamIdPara(Nan::To<v8::String>(info[1]).ToLocalChecked());
+  std::string streamId = std::string(*streamIdPara);
+
+  Nan::Utf8String ownerIdPara(Nan::To<v8::String>(info[2]).ToLocalChecked());
+  std::string ownerId = std::string(*ownerIdPara);
+
+  me->addInput(src, streamId, ownerId);
+}
+
+NAN_METHOD(AudioRanker::removeInput) {
+  AudioRanker* obj = Nan::ObjectWrap::Unwrap<AudioRanker>(info.Holder());
+  owt_base::AudioRanker* me = obj->me;
+
+  Nan::Utf8String streamIdPara(Nan::To<v8::String>(info[0]).ToLocalChecked());
+  std::string streamId = std::string(*streamIdPara);
+
+  me->removeInput(streamId);
+}
+
+void AudioRanker::onRankChange(std::vector<std::pair<std::string, std::string>> updates) {
+  boost::mutex::scoped_lock lock(mutex);
+  std::ostringstream jsonChange;
+  /*
+   * The JS callback json
+   * [
+   *    ["streamID0", "ownerID0"],
+   *    ["streamID1", "ownerID1"],
+   *    ......
+   *    ["streamIDn", "ownerIDn"]
+   * ]
+   */
+  jsonChange.str("");
+  jsonChange << "[";
+  for (size_t i = 0; i + 1 < updates.size(); i++) {
```

Comment:
Finish this in a single loop instead of leaving last element there merely because of different formatting.